compiler/util/src/org/jetbrains/kotlin/utils/SmartHashMap.kt

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.utils
+
+const val ARRAY_UNTIL_SIZE = 10
+
+/**
+ * [SmartHashMap] is a Map implementation that uses reference identity for keys.
+ * It uses 2 arrays to store keys & values until the number of entries stored is larger than 10.
+ * At that point it switches to using a HashMap and stays that way until [clear] is called.
+ * It does not auto convert back if the number of entries decreases through [remove].
+ */
+class SmartHashMap<K, V>() : MutableMap<K, V> {
+
+    private var keysArray: MutableList<K>? = arrayListOf()
+    private var valuesArray: MutableList<V>? = arrayListOf()
+    private var largeMap: HashMap<K, V>? = null
+
+    override val entries: MutableSet<MutableMap.MutableEntry<K, V>>
+        get() {
+            return keysArray?.let {
+                val ms = mutableSetOf<MutableMap.MutableEntry<K, V>>()
+                for (i in 0 until it.size) {
+                    ms.add(Entry(it[i], valuesArray!![i]))
+                }
+                return ms
+            } ?: largeMap!!.entries
+        }
+
+
+    override val keys: MutableSet<K>
+        get() {
+            return keysArray?.let {
+                val ms = mutableSetOf<K>()
+                ms.addAll(it)
+                return ms
+
+            } ?: largeMap!!.keys
+        }
+
+    override val size: Int
+        get() = keysArray?.size ?: largeMap!!.size
+
+    override val values: MutableCollection<V>
+        get() {
+            return valuesArray?.let {
+                val ms = mutableSetOf<V>()
+                ms.addAll(it)
+                return ms
+
+            } ?: largeMap!!.values
+        }
+
+    override fun clear() {
+        largeMap = null
+        keysArray = arrayListOf()
+        valuesArray = arrayListOf()
+    }
+
+    override fun containsKey(key: K): Boolean = keysArray?.contains(key) ?: largeMap!!.containsKey(key)
+
+
+    override fun containsValue(value: V): Boolean = valuesArray?.contains(value) ?: largeMap!!.containsValue(value)
+
+    override fun get(key: K): V? {
+        return keysArray?.let {
+
+            for (i in 0 until it.size) {
+                if (it[i] === key) {
```

Comment:
The underlying HashMap and MutableList compare with `.equals()` rather than identity so most of SmartHashMap still use equality except here and below.

Reply:
cleaned the whole file for this pattern, using IdentityHashMap as Mark suggested and replaced the calls to contains with any{x===y}